### PR TITLE
Apply statsmodels-based ARIMA/VARIMA to new TS

### LINKED ISSUE #920: Apply ARIMA to a new time series
It seems that statsmodels allows applying a fitted ARIMA model on another series: https://www.statsmodels.org/dev/generated/statsmodels.tsa.arima.model.ARIMAResults.apply.html

We should use this to accept an optional `series` argument in the predict function. That would probably require a new super-class for the ARIMA model.

### PR DESCRIPTION
**Pull request by piaz97**
Fixes #920 .

### Summary
ARIMA and VARIMA models wrapping statsmodels implementations now support the prediction new TS (in addition to the one used during training). 

### Other Information
- New base class added, extending `DualCovariatesForecastingModel`, and adding the support for a `series` parameter in the `predict()` method. Not satisfied with the new base class name, `StatsmodelsDualCovariatesForecastingModel`, any better idea would be welcomed.
- Adapted both `ARIMA` and `VARIMA` models.
- Added tests.

### CODE REVIEW

File: darts/models/forecasting/forecasting_model.py [hunk 1]
```diff
@@ -1228,3 +1228,129 @@ def _predict_wrapper(
         return self.predict(
             n, future_covariates=future_covariates, num_samples=num_samples
         )
+
+
+class StatsmodelsDualCovariatesForecastingModel(DualCovariatesForecastingModel, ABC):
+    """The base class for the forecasting models that are not global, but support future covariates, and can
+    additionally be applied to new data unrelated to the original series used for fitting the model. Currently,
+    all the derives classes wrap statsmodels models.
```
Comment (brunnedu) [inline darts/models/forecasting/forecasting_model.py:1236]:
```suggestion
    all the derived classes wrap statsmodels models.
```

File: darts/models/forecasting/forecasting_model.py [hunk 2]
```diff
@@ -1228,3 +1228,129 @@ def _predict_wrapper(
...
+    def _predict(
+        self,
+        n: int,
+        series: Optional[TimeSeries] = None,
+        historic_future_covariates: Optional[TimeSeries] = None,
+        future_covariates: Optional[TimeSeries] = None,
+        num_samples: int = 1,
+    ) -> TimeSeries:
+        """Forecasts values for a certain number of time steps after the end of the series.
+        ExtendedDualCovariatesModels must implement the predict logic in this method.
```
Comment (brunnedu) [inline darts/models/forecasting/forecasting_model.py:1339]:
```suggestion
        StatsmodelsDualCovariatesForecastingModels must implement the predict logic in this method.
```

**Review by brunnedu — COMMENTED** [review]
Cool addition to darts! 🚀 Left a few minor comments

File: darts/models/forecasting/arima.py [hunk 3]
```diff
@@ -107,6 +125,15 @@ def _predict(
                 exog=future_covariates.values() if future_covariates else None,
             )
 
+        # restoring statsmodels results object state
```
Comment (hrzn) [inline darts/models/forecasting/arima.py:128]: Why is this needed?

File: darts/models/forecasting/varima.py [hunk 4]
```diff
@@ -92,15 +102,57 @@ def _fit(
...
+            # updating statsmodels results object state
+            self.model = self.model.apply(
+                series.values(),
+                exog=historic_future_covariates.values()
+                if historic_future_covariates
+                else None,
+            )
+
+        # forecast before restoring the training state
         forecast = self.model.forecast(
```
Comment (hrzn) [inline darts/models/forecasting/varima.py:139]: While we are at it, is there a chance we could use something like `simulate()` as we use for ARIMA here, to make VARIMA probabilistic too?

**Review by hrzn — COMMENTED** [review]
Looks good! Only a few minor comments and a suggestion name for the new class

File: darts/models/forecasting/forecasting_model.py [hunk 5]
```diff
@@ -1228,3 +1228,129 @@ def _predict_wrapper(
         return self.predict(
             n, future_covariates=future_covariates, num_samples=num_samples
         )
+
+
+class StatsmodelsDualCovariatesForecastingModel(DualCovariatesForecastingModel, ABC):
```
[the remark this answers is not in the document]
Comment (piaz97) [reply darts/models/forecasting/forecasting_model.py:1233]: Thanks! Sounds better, I'll go for this one 👍🏼

File: darts/models/forecasting/arima.py (hunk 3 above)
Comment (piaz97) [reply darts/models/forecasting/arima.py:128]: The default behavior of the `predict()` method, without specifying any `series`, forecasts the next `n` steps of the training series. To preserve this behavior also after calling `predict` with a different TS (which changes the state of the underlying `ARIMAResults` object calling `apply()`), I call `apply` again with the training data.

File: darts/models/forecasting/arima.py [hunk 6]
```diff
@@ -93,7 +100,18 @@ def _predict(
...
-        super()._predict(n, future_covariates, num_samples)
+        super()._predict(
+            n, series, historic_future_covariates, future_covariates, num_samples
+        )
+
+        # updating statsmodels results object state with the new ts and covariates
+        if series is not None:
+            self.model = self.model.apply(
+                series.values(),
+                exog=historic_future_covariates.values()
```
Comment (hrzn) [inline darts/models/forecasting/arima.py:111]:
```suggestion
                exog=historic_future_covariates.values(copy=False)
```

File: darts/models/forecasting/varima.py [hunk 7]
```diff
@@ -92,27 +102,95 @@ def _fit(
...
+
+            # if the series is differentiated, the new len will be = len - 1, we have to adjust the future covariates
+            if historic_future_covariates and self.d > 0:
+                historic_future_covariates = historic_future_covariates.slice_intersect(
+                    series
+                )
+
+            # updating statsmodels results object state
+            self.model = self.model.apply(
+                series.values(),
```
Comment (hrzn) [inline darts/models/forecasting/varima.py:141]:
```suggestion
                series.values(copy=False),
```

File: darts/models/forecasting/varima.py [hunk 8]
```diff
@@ -92,27 +102,95 @@ def _fit(
...
+                exog=historic_future_covariates.values()
```
Comment (hrzn) [inline darts/models/forecasting/varima.py:142]:
```suggestion
                exog=historic_future_covariates.values(copy=False)
```

File: darts/models/forecasting/varima.py [hunk 9]
```diff
@@ -92,27 +102,95 @@ def _fit(
...
+                if historic_future_covariates
+                else None,
+            )
+
+        # forecast before restoring the training state
+        if num_samples == 1:
+            forecast = self.model.forecast(
+                steps=n, exog=future_covariates.values() if future_covariates else None
```
Comment (hrzn) [inline darts/models/forecasting/varima.py:150]:
```suggestion
                steps=n, exog=future_covariates.values(copy=False) if future_covariates else None
```

File: darts/models/forecasting/varima.py [hunk 10]
```diff
@@ -92,27 +102,95 @@ def _fit(
...
+            )
+        else:
+            forecast = self.model.simulate(
+                nsimulations=n,
+                repetitions=num_samples,
+                initial_state=self.model.states.predicted[-1, :],
+                exog=future_covariates.values() if future_covariates else None,
```
Comment (hrzn) [inline darts/models/forecasting/varima.py:157]:
```suggestion
                exog=future_covariates.values(copy=False) if future_covariates else None,
```

File: darts/models/forecasting/varima.py [hunk 11]
```diff
@@ -92,27 +102,95 @@ def _fit(
...
+            )
+
         forecast = self._invert_transformation(forecast)
+
+        # restoring statsmodels results object state and last values
+        if series is not None:
+            self.model = self.model.apply(
+                self._orig_training_series.values(),
```
Comment (hrzn) [inline darts/models/forecasting/varima.py:165]:
```suggestion
                self._orig_training_series.values(copy=False),
```

File: darts/models/forecasting/varima.py [hunk 12]
```diff
@@ -92,27 +102,95 @@ def _fit(
...
+                exog=self.training_historic_future_covariates.values()
```
Comment (hrzn) [inline darts/models/forecasting/varima.py:166]:
```suggestion
                exog=self.training_historic_future_covariates.values(copy=False)
```

**Review by hrzn — APPROVED** [review]
Looks good! I've identified a few opportunities for (small) performance improvements using `copy=False` - there might be more that I missed! 

Before merging, would you agree to check that we do not unnecessarily prevent historical forecasts with `retrain=False` anymore? I think all it takes will likely be to override the method `_supports_non_retrainable_historical_forecasts()` in `TransferrableDualCovariatesForecastingModel` to return `True` instead of `False`.

**Comment by hrzn** [comment]
@piaz97 the tests are failing because `ignore_time_axes` was renamed into `ignore_time_axis`. Could you rename here too?

**Comment by piaz97** [comment]
I added the suggestions from the previous comments, added support for backtesting with` retrain=False`, and minor fixes. Should be ready to be merged
